akka-actor-typed/src/main/scala/akka/actor/typed/delivery/internal/ConsumerControllerImpl.scala

```diff
@@ -0,0 +1,467 @@
+        receiveStart(s, start, newState => resending(newState))
+
+      case ConsumerTerminated(c) =>
+        receiveConsumerTerminated(c)
+
+      case reg: RegisterToProducerController[A] =>
+        receiveRegisterToProducerController(s, reg, newState => active(newState))
+
+      case _: UnsealedInternalCommand =>
+        Behaviors.unhandled
+    }
+  }
+
+  private def deliver(s: State[A], seqMsg: SequencedMessage[A]): Behavior[InternalCommand] = {
+    context.log.info("from producer [{}], deliver [{}] to consumer", seqMsg.producerId, seqMsg.seqNr)
+    s.consumer ! Delivery(seqMsg.producerId, seqMsg.seqNr, seqMsg.msg, context.self)
+    waitingForConfirmation(s, seqMsg)
+  }
+
+  // The message has been delivered to the consumer and it is now waiting for Confirmed from
+  // the consumer. New SequencedMessage from the ProducerController will be stashed.
+  private def waitingForConfirmation(s: State[A], seqMsg: SequencedMessage[A]): Behavior[InternalCommand] = {
+    Behaviors.receiveMessage {
+      case Confirmed(seqNr) =>
+        val expectedSeqNr = s.receivedSeqNr
+        if (seqNr > expectedSeqNr) {
+          throw new IllegalStateException(
+            s"Expected confirmation of seqNr [$expectedSeqNr], but received higher [$seqNr]")
+        } else if (seqNr != expectedSeqNr) {
+          context.log.info(
+            "Expected confirmation of seqNr [{}] but received [{}]. Perhaps the consumer was restarted.",
+            expectedSeqNr,
+            seqNr)
+        }
+        context.log.info("Confirmed [{}], stashed size [{}]", seqNr, stashBuffer.size)
+
+        val newRequestedSeqNr =
+          if (seqMsg.first) {
+            // confirm the first message immediately to cancel resending of first
+            val newRequestedSeqNr = seqNr - 1 + flowControlWindow
+            context.log.info("Request after first [{}]", newRequestedSeqNr)
+            s.producer ! Request(confirmedSeqNr = seqNr, newRequestedSeqNr, resendLost, viaTimeout = false)
+            newRequestedSeqNr
+          } else if ((s.requestedSeqNr - seqNr) == flowControlWindow / 2) {
+            val newRequestedSeqNr = s.requestedSeqNr + flowControlWindow / 2
+            context.log.info("Request [{}]", newRequestedSeqNr)
+            s.producer ! Request(confirmedSeqNr = seqNr, newRequestedSeqNr, resendLost, viaTimeout = false)
+            startRetryTimer() // reset interval since Request was just sent
+            newRequestedSeqNr
+          } else {
+            if (seqMsg.ack) {
+              context.log.info("Ack [{}]", seqNr)
+              s.producer ! Ack(confirmedSeqNr = seqNr)
+            }
+
+            s.requestedSeqNr
+          }
+
+        // FIXME can we use unstashOne instead of all?
+        if (stashBuffer.nonEmpty)
+          context.log.info("Unstash [{}]", stashBuffer.size)
+        stashBuffer.unstashAll(active(s.copy(confirmedSeqNr = seqNr, requestedSeqNr = newRequestedSeqNr)))
+
+      case seqMsg: SequencedMessage[A] =>
```

Comment:
I don't think there will be lots of duplicates since messages are not pushed from the producer, apart from the first message. Anyway, I added a dedup if the incoming is the same as we are awaiting